File: server/agent/apply.js
'use strict';

const linear = require('../linear');
const { CONFIG } = require('../config');
const { normalizeTshirtSize } = require('./schema');

/**
 * Resolve the AI task-label id once per apply run so every created issue can be
 * stamped with it (Step 2: software-design issues are marked with the AI label,
 * which the coder board monitor picks up in Step 3). Best-effort: on failure we
 * return null and issues are created unlabeled rather than aborting the plan.
 */
async function resolveTaskLabelId(apiKey, step) {
  try {
    const label = await linear.getOrCreateIssueLabel(apiKey, CONFIG.CODER.taskLabel);
    return label.id;
  } catch (err) {
    step(`Could not resolve "${CONFIG.CODER.taskLabel}" issue label: ${errMsg(err)}`, 'warn');
    return null;
  }
}

/** Append a labelled criteria block to a description (no-op if empty). */
function withCriteria(base, label, criteria) {
  const b = base || '';
  if (!criteria) return b;
  return `${b ? `${b}\n\n` : ''}**${label}:** ${criteria}`;
}

/** The model-routing label for a T-shirt size: XS → local, everything larger → hosted. */
function modelLabelForSize(size) {
  return size === String(CONFIG.CODER.localSize).toUpperCase() ? CONFIG.CODER.localModelLabel : CONFIG.CODER.hostedModelLabel;
}

/**
 * Memoized issue-label resolver: getOrCreate each distinct label name once per
 * apply run and cache its id. Best-effort — a failure caches null so the issue is
 * created without that label rather than aborting the whole plan.
 */
function makeLabelResolver(apiKey, step) {
  const cache = new Map();
  return async (name) => {
    if (!name) return null;
    if (cache.has(name)) return cache.get(name);
    try {
      const label = await linear.getOrCreateIssueLabel(apiKey, name);
      cache.set(name, label.id);
      return label.id;
    } catch (err) {
      step(`Could not resolve "${name}" issue label: ${errMsg(err)}`, 'warn');
      cache.set(name, null);
      return null;
    }
  };
}

/**
 * Memoized resolver for the model-routing label (local/hosted), resolved as a
 * member of the "Models" label group so Linear shows it as a single-select
 * dropdown on the issue. Best-effort: on failure it falls back to an ungrouped
 * label, then to null, so a labelling hiccup never aborts issue creation.
 */
function makeModelLabelResolver(apiKey, step) {
  const cache = new Map();
  const groupName = CONFIG.CODER.modelLabelGroup;
  return async (name) => {
    if (!name) return null;
    if (cache.has(name)) return cache.get(name);
    let id = null;
    try {
      id = (await linear.getOrCreateGroupedIssueLabel(apiKey, groupName, name)).id;
    } catch (err) {
      step(`Could not group "${name}" under "${groupName}": ${errMsg(err)}`, 'warn');
      try {
        id = (await linear.getOrCreateIssueLabel(apiKey, name)).id;
      } catch (err2) {
        step(`Could not resolve "${name}" issue label: ${errMsg(err2)}`, 'warn');
      }
    }
    cache.set(name, id);
    return id;
  };
}

/** Label ids to stamp on an issue: task label + T-shirt size + model-routing label. */
async function issueLabelIds(resolveLabel, resolveModelLabel, taskLabelId, size) {
  const sizeName = normalizeTshirtSize(size); // always a valid XS|S|M|L|XL
  const [sizeId, modelId] = await Promise.all([
    resolveLabel(sizeName),
    resolveModelLabel(modelLabelForSize(sizeName)),
  ]);
  return [taskLabelId, sizeId, modelId].filter(Boolean);
}

/**
 * Apply a validated, normalized enrichment plan to a Linear project using the
 * stored Linear token. Writes are deterministic (not performed by the LLM):
 * the LLM proposes, the server disposes — a hard guardrail per the
 * ai-prompt-injection checklist (no autonomous destructive tool calls).
 *
 * Resilient: a single milestone/issue/dependency failure is recorded as a
 * warning and does not abort the rest of the plan.
 */
async function applyPlan(apiKey, { project, plan, assumedRole, config, onStep }) {
  const step = typeof onStep === 'function' ? onStep : () => {};
  const warnings = [];
  const summary = { milestonesCreated: 0, issuesCreated: 0, dependenciesCreated: 0, warnings };

  const { team } = await linear.getProjectTeam(apiKey, project.id);
  step(`Applying plan to Linear (team ${team.key || team.name}).`);
  const taskLabelId = config.createIssues ? await resolveTaskLabelId(apiKey, step) : null;
  const resolveLabel = makeLabelResolver(apiKey, step);
  const resolveModelLabel = makeModelLabelResolver(apiKey, step);

  // 1. Assign the assumed role as project lead (claims the open project).
  if (config.autoAssignLead && assumedRole && assumedRole.id) {
    try {
      await linear.setProjectLead(apiKey, project.id, assumedRole.id);
      step(`Assigned lead: ${assumedRole.name}.`);
    } catch (err) {
      warnings.push(`Lead assignment failed: ${errMsg(err)}`);
      step(`Lead assignment failed: ${errMsg(err)}`, 'warn');
    }
  }

  // 2. Enrich the project description.
  try {
    await linear.updateProjectDescription(apiKey, project.id, plan.description);
    step('Updated project description.');
  } catch (err) {
    warnings.push(`Description update failed: ${errMsg(err)}`);
    step(`Description update failed: ${errMsg(err)}`, 'warn');
  }

  // 3. Create milestones. Linear milestones only carry a targetDate, so the
  //    start date is preserved in the description as an explicit timeline line.
  const issueIdMatrix = []; // [milestoneIndex][issueIndex] -> issueId
  for (const milestone of plan.milestones) {
    const timelineNote = `**Timeline:** ${milestone.startDate} → ${milestone.targetDate}`;
    let description = milestone.description ? `${timelineNote}\n\n${milestone.description}` : timelineNote;
    description = withCriteria(description, 'Evaluation criteria', milestone.evaluationCriteria);

    let createdMilestone = null;
    try {
      createdMilestone = await linear.createMilestone(apiKey, {
        projectId: project.id,
        name: milestone.name,
        description,
        targetDate: milestone.targetDate,
      });
      summary.milestonesCreated += 1;
      step(`Created milestone: ${milestone.name} (${milestone.startDate} → ${milestone.targetDate}).`);
    } catch (err) {
      warnings.push(`Milestone "${milestone.name}" failed: ${errMsg(err)}`);
      step(`Milestone "${milestone.name}" failed: ${errMsg(err)}`, 'warn');
      issueIdMatrix.push([]);
      continue;
    }

    // 4. Create issues for this milestone.
    const issueIds = [];
    if (config.createIssues) {
      for (const issue of milestone.issues) {
        try {
          const labelIds = await issueLabelIds(resolveLabel, resolveModelLabel, taskLabelId, issue.tshirtSize);
          const created = await linear.createIssue(apiKey, {
            teamId: team.id,
            projectId: project.id,
            projectMilestoneId: createdMilestone.id,
            title: issue.title,
            description: withCriteria(issue.description, 'Acceptance criteria', issue.evaluationCriteria),
            priority: issue.priority,
            labelIds: labelIds.length ? labelIds : undefined,
          });
          issueIds.push(created.id);
          summary.issuesCreated += 1;
        } catch (err) {
          issueIds.push(null);
          warnings.push(`Issue "${issue.title}" failed: ${errMsg(err)}`);
          step(`Issue "${issue.title}" failed: ${errMsg(err)}`, 'warn');
        }
      }
      const ok = issueIds.filter(Boolean).length;
      if (ok) step(`Created ${ok} issue(s) under "${milestone.name}".`);
    }
    issueIdMatrix.push(issueIds);
  }

  // 5. Create dependencies (from blocks to). Indices were already validated in
  //    normalizePlan; we re-check the resolved IDs before writing.
  if (config.addDependencies && config.createIssues) {
    for (const dep of plan.dependencies) {
      const fromId = safeAt(issueIdMatrix, dep.fromMilestone, dep.fromIssue);
      const toId = safeAt(issueIdMatrix, dep.toMilestone, dep.toIssue);
      if (!fromId || !toId || fromId === toId) continue;
      try {
        await linear.createIssueRelation(apiKey, {
          issueId: fromId,
          relatedIssueId: toId,
          type: 'blocks',
        });
        summary.dependenciesCreated += 1;
      } catch (err) {
        warnings.push(`Dependency failed: ${errMsg(err)}`);
        step(`Dependency failed: ${errMsg(err)}`, 'warn');
      }
    }
    if (summary.dependenciesCreated) step(`Created ${summary.dependenciesCreated} dependency link(s).`);
  }

  return summary;
}

/**
 * Resume path: create issues for EXISTING milestones that currently have none.
 * `generated` holds tasks per milestone (same order as `milestones`); matched by
 * index, then by name as a fallback.
 */
async function applyIssuesForMilestones(apiKey, { project, milestones, generated, config, onStep }) {
  const step = typeof onStep === 'function' ? onStep : () => {};
  const warnings = [];
  const summary = { milestonesCreated: 0, issuesCreated: 0, dependenciesCreated: 0, warnings, resumed: true };
  if (!config.createIssues) return summary;

  const { team } = await linear.getProjectTeam(apiKey, project.id);
  step(`Creating tasks for ${milestones.length} milestone(s) (team ${team.key || team.name}).`);
  const taskLabelId = await resolveTaskLabelId(apiKey, step);
  const resolveLabel = makeLabelResolver(apiKey, step);
  const resolveModelLabel = makeModelLabelResolver(apiKey, step);

  for (let i = 0; i < milestones.length; i += 1) {
    const milestone = milestones[i];
    const match = generated[i] || generated.find((g) => (g.name || '').toLowerCase() === (milestone.name || '').toLowerCase());
    const issues = (match && match.issues) || [];

    // Add the milestone's evaluation criteria to its description (if any).
    if (match && match.evaluationCriteria) {
      try {
        await linear.updateMilestone(apiKey, {
          id: milestone.id,
          description: withCriteria(milestone.description, 'Evaluation criteria', match.evaluationCriteria),
        });
        step(`Added evaluation criteria to "${milestone.name}".`);
      } catch (err) {
        warnings.push(`Milestone criteria update failed: ${errMsg(err)}`);
      }
    }

    let created = 0;
    for (const issue of issues) {
      try {
        const labelIds = await issueLabelIds(resolveLabel, resolveModelLabel, taskLabelId, issue.tshirtSize);
        await linear.createIssue(apiKey, {
          teamId: team.id,
          projectId: project.id,
          projectMilestoneId: milestone.id,
          title: issue.title,
          description: withCriteria(issue.description, 'Acceptance criteria', issue.evaluationCriteria),
          priority: issue.priority,
          labelIds: labelIds.length ? labelIds : undefined,
        });
        created += 1;
        summary.issuesCreated += 1;
      } catch (err) {
        warnings.push(`Issue "${issue.title}" failed: ${errMsg(err)}`);
        step(`Issue "${issue.title}" failed: ${errMsg(err)}`, 'warn');
      }
    }
    if (created) step(`Created ${created} task(s) under "${milestone.name}".`);
  }
  return summary;
}

/** Mark a project complete: switch its label to `aidone` (replacing others). */
async function applyAidone(apiKey, { project, onStep }) {
  const step = typeof onStep === 'function' ? onStep : () => {};
  try {
    const label = await linear.getOrCreateProjectLabel(apiKey, 'aidone');
    await linear.setProjectLabels(apiKey, project.id, [label.id]);
    step('Set project label to "aidone".');
  } catch (err) {
    step(`aidone label failed: ${errMsg(err)}`, 'warn');
  }
}

/**
 * Mark a project as PLANNED: switch its label to `aiplanned` (replacing the enrich
 * label). This both drops the project out of the planning set and signals the
 * coding flow to start working its tasks in dependency order.
 */
async function applyAiplanned(apiKey, { project, onStep }) {
  const step = typeof onStep === 'function' ? onStep : () => {};
  try {
    const label = await linear.getOrCreateProjectLabel(apiKey, 'aiplanned');
    await linear.setProjectLabels(apiKey, project.id, [label.id]);
    step('Set project label to "aiplanned".');
  } catch (err) {
    step(`aiplanned label failed: ${errMsg(err)}`, 'warn');
  }
}

/**
 * Handle a project the business-owner agent judged NOT viable: append a note to
 * the project description and switch its label to `aifail` (replacing existing
 * labels so it leaves the enrichment set and is not retried).
 */
async function applyAifail(apiKey, { project, reason, onStep }) {
  const step = typeof onStep === 'function' ? onStep : () => {};
  const note = `\n\n---\n**AI viability check — not a fit for a software-driven solution.**\nReason: ${reason}`;
  const nextDescription = `${project.description || ''}${note}`.trim();

  try {
    await linear.updateProjectDescription(apiKey, project.id, nextDescription);
    step('Added viability note to project description.');
  } catch (err) {
    step(`Description note failed: ${errMsg(err)}`, 'warn');
  }

  try {
    const label = await linear.getOrCreateProjectLabel(apiKey, 'aifail');
    await linear.setProjectLabels(apiKey, project.id, [label.id]);
    step('Set project label to "aifail".');
  } catch (err) {
    step(`Label update failed: ${errMsg(err)}`, 'warn');
  }

  return { aifail: true, reason, milestonesCreated: 0, issuesCreated: 0, dependenciesCreated: 0, warnings: [] };
}

/* --------------------- Coder issue state transitions -------------------- */

/**
 * Move a coder task to "In Progress" (a `started` state) before the agent runs.
 * Idempotent: an issue already started/completed is left as-is. Returns the
 * resulting state (or null if the team has no started state).
 */
async function startIssue(apiKey, { issueId, onStep }) {
  const step = typeof onStep === 'function' ? onStep : () => {};
  const detail = await linear.getIssueDetail(apiKey, issueId);
  const type = detail.state && detail.state.type;
  if (type === 'started' || type === 'completed' || type === 'canceled') {
    return detail.state || null; // already in/after progress — nothing to do
  }
  const target = linear.pickStateByType(await linear.getTeamStates(apiKey, detail.team.id), 'started', 'In Progress');
  if (!target) {
    step('No "In Progress" workflow state on this team; leaving state unchanged.', 'warn');
    return detail.state || null;
  }
  await linear.updateIssue(apiKey, issueId, { stateId: target.id });
  step(`Moved issue to "${target.name}".`);
  return target;
}

/**
 * Finish a coder task per the agent's verdict: move it to Done (a `completed`
 * state) and stamp the outcome label on the ISSUE — `aidone` when completed,
 * `aifail` when insufficient — creating the label if missing and appending it to
 * the issue's existing labels. An insufficient reason is posted as a comment.
 */
async function finishIssue(apiKey, { issueId, outcome, reason, onStep }) {
  const step = typeof onStep === 'function' ? onStep : () => {};
  const labelName = outcome === 'completed' ? 'aidone' : 'aifail';
  const detail = await linear.getIssueDetail(apiKey, issueId);

  // Resolve (create-if-missing) the outcome label, appended to existing labels.
  let labelIds;
  try {
    const label = await linear.getOrCreateIssueLabel(apiKey, labelName);
    const current = (detail.labels && detail.labels.nodes ? detail.labels.nodes : []).map((l) => l.id);
    labelIds = [...new Set([...current, label.id])];
  } catch (err) {
    step(`Could not resolve "${labelName}" issue label: ${errMsg(err)}`, 'warn');
  }

  const done = linear.pickStateByType(await linear.getTeamStates(apiKey, detail.team.id), 'completed', 'Done');
  const input = {};
  if (done) input.stateId = done.id;
  if (labelIds) input.labelIds = labelIds;
  if (Object.keys(input).length) await linear.updateIssue(apiKey, issueId, input);
  step(`Marked issue ${done ? `"${done.name}"` : '(no Done state found)'} + label "${labelName}".`);

  // Record the reason a task was judged insufficient (for the human triaging it).
  if (outcome !== 'completed' && reason) {
    try {
      await linear.createComment(apiKey, { issueId, body: `**AI coder — insufficient to complete.**\n\n${reason}` });
      step('Posted insufficiency reason as a comment.');
    } catch (err) {
      step(`Comment failed: ${errMsg(err)}`, 'warn');
    }
  }
  return { labelName, done: Boolean(done) };
}

function safeAt(matrix, i, j) {
  return matrix[i] && matrix[i][j] ? matrix[i][j] : null;
}

function errMsg(err) {
  return err && err.message ? err.message : String(err);
}

module.exports = { applyPlan, applyIssuesForMilestones, applyAidone, applyAiplanned, applyAifail, startIssue, finishIssue };

File: ai_fleet/agent/apply.py
"""Deterministic server-side application of enrichment plans to Linear.

Port of ``packages/shared/src/agent/apply.js``. The LLM proposes, the server
disposes: every Linear write here is performed by our own code (not an
autonomous tool call), a hard guardrail per the ai-prompt-injection checklist.

Resilient by design: a single milestone/issue/dependency failure is recorded as
a warning and does not abort the rest of the plan.
"""

from __future__ import annotations

import inspect

from ai_fleet import linear
from ai_fleet.config import CONFIG
from ai_fleet.agent.schema import normalize_tshirt_size


def _err_msg(err) -> str:
    """JS ``err && err.message ? err.message : String(err)``."""
    msg = getattr(err, "message", None)
    return msg if msg else str(err)


def _cfg(config, key):
    """Read a config field, tolerating either a dict (store contract, camelCase
    keys) or an attribute-style object."""
    if config is None:
        return None
    if isinstance(config, dict):
        return config.get(key)
    return getattr(config, key, None)


def _make_step(on_step):
    """Wrap the optional ``on_step`` callback into a ``step(message, level=None)``
    function. Mirrors the JS ``const step = typeof onStep === 'function' ? onStep
    : () => {}`` — a no-op when no callback is provided.

    The JS callback is called with either ``(message)`` or ``(message, 'warn')``
    and extra args are ignored. Python is stricter about arity, so we only pass
    the level when the callback can accept a second positional argument.
    """
    if not callable(on_step):
        return lambda message, level=None: None

    accepts_level = True
    try:
        params = list(inspect.signature(on_step).parameters.values())
        positional = [
            p for p in params
            if p.kind in (p.POSITIONAL_ONLY, p.POSITIONAL_OR_KEYWORD)
        ]
        has_var_positional = any(p.kind == p.VAR_POSITIONAL for p in params)
        accepts_level = has_var_positional or len(positional) >= 2
    except (ValueError, TypeError):
        accepts_level = True

    def step(message, level=None):
        if level is not None and accepts_level:
            on_step(message, level)
        else:
            on_step(message)

    return step


async def _resolve_task_label_id(api_key, step):
    """Resolve the AI task-label id once per apply run so every created issue can
    be stamped with it (Step 2). Best-effort: on failure returns None and issues
    are created unlabeled rather than aborting the plan."""
    try:
        label = await linear.get_or_create_issue_label(api_key, CONFIG.CODER.taskLabel)
        return label["id"]
    except Exception as err:
        step(f'Could not resolve "{CONFIG.CODER.taskLabel}" issue label: {_err_msg(err)}', "warn")
        return None


def _with_criteria(base, label, criteria):
    """Append a labelled criteria block to a description (no-op if empty)."""
    b = base or ""
    if not criteria:
        return b
    prefix = f"{b}\n\n" if b else ""
    return f"{prefix}**{label}:** {criteria}"


def _model_label_for_size(size):
    """The model-routing label for a T-shirt size: XS → local, larger → hosted."""
    if size == str(CONFIG.CODER.localSize).upper():
        return CONFIG.CODER.localModelLabel
    return CONFIG.CODER.hostedModelLabel


def _make_label_resolver(api_key, step):
    """Memoized issue-label resolver: getOrCreate each distinct label name once
    per apply run and cache its id. Best-effort — a failure caches None so the
    issue is created without that label rather than aborting the whole plan."""
    cache: dict = {}

    async def resolve(name):
        if not name:
            return None
        if name in cache:
            return cache[name]
        try:
            label = await linear.get_or_create_issue_label(api_key, name)
            cache[name] = label["id"]
            return label["id"]
        except Exception as err:
            step(f'Could not resolve "{name}" issue label: {_err_msg(err)}', "warn")
            cache[name] = None
            return None

    return resolve


def _make_model_label_resolver(api_key, step):
    """Memoized resolver for the model-routing label (local/hosted), resolved as
    a member of the "Models" label group so Linear shows it as a single-select
    dropdown. Best-effort: on failure it falls back to an ungrouped label, then
    to None, so a labelling hiccup never aborts issue creation."""
    cache: dict = {}
    group_name = CONFIG.CODER.modelLabelGroup

    async def resolve(name):
        if not name:
            return None
        if name in cache:
            return cache[name]
        label_id = None
        try:
            label_id = (await linear.get_or_create_grouped_issue_label(api_key, group_name, name))["id"]
        except Exception as err:
            step(f'Could not group "{name}" under "{group_name}": {_err_msg(err)}', "warn")
            try:
                label_id = (await linear.get_or_create_issue_label(api_key, name))["id"]
            except Exception as err2:
                step(f'Could not resolve "{name}" issue label: {_err_msg(err2)}', "warn")
        cache[name] = label_id
        return label_id

    return resolve


async def _issue_label_ids(resolve_label, resolve_model_label, task_label_id, size):
    """Label ids to stamp on an issue: task label + T-shirt size + model label."""
    size_name = normalize_tshirt_size(size)  # always a valid XS|S|M|L|XL
    size_id = await resolve_label(size_name)
    model_id = await resolve_model_label(_model_label_for_size(size_name))
    return [label_id for label_id in (task_label_id, size_id, model_id) if label_id]


async def apply_plan(api_key, *, project, plan, assumed_role, config, on_step=None):
    """Apply a validated, normalized enrichment plan to a Linear project using the
    stored Linear token. Writes are deterministic (not performed by the LLM)."""
    step = _make_step(on_step)
    warnings: list = []
    summary = {"milestonesCreated": 0, "issuesCreated": 0, "dependenciesCreated": 0, "warnings": warnings}

    team = (await linear.get_project_team(api_key, project["id"]))["team"]
    step(f"Applying plan to Linear (team {team.get('key') or team.get('name')}).")
    task_label_id = await _resolve_task_label_id(api_key, step) if _cfg(config, "createIssues") else None
    resolve_label = _make_label_resolver(api_key, step)
    resolve_model_label = _make_model_label_resolver(api_key, step)

    # 1. Assign the assumed role as project lead (claims the open project).
    if _cfg(config, "autoAssignLead") and assumed_role and assumed_role.get("id"):
        try:
            await linear.set_project_lead(api_key, project["id"], assumed_role["id"])
            step(f"Assigned lead: {assumed_role.get('name')}.")
        except Exception as err:
            warnings.append(f"Lead assignment failed: {_err_msg(err)}")
            step(f"Lead assignment failed: {_err_msg(err)}", "warn")

    # 2. Enrich the project description.
    try:
        await linear.update_project_description(api_key, project["id"], plan["description"])
        step("Updated project description.")
    except Exception as err:
        warnings.append(f"Description update failed: {_err_msg(err)}")
        step(f"Description update failed: {_err_msg(err)}", "warn")

    # 3. Create milestones. Linear milestones only carry a targetDate, so the
    #    start date is preserved in the description as an explicit timeline line.
    issue_id_matrix: list = []  # [milestoneIndex][issueIndex] -> issueId
    for milestone in plan["milestones"]:
        name = milestone.get("name")
        start_date = milestone.get("startDate")
        target_date = milestone.get("targetDate")
        timeline_note = f"**Timeline:** {start_date} → {target_date}"
        description = f"{timeline_note}\n\n{milestone['description']}" if milestone.get("description") else timeline_note
        description = _with_criteria(description, "Evaluation criteria", milestone.get("evaluationCriteria"))

        try:
            created_milestone = await linear.create_milestone(
                api_key,
                project["id"],
                name,
                description=description,
                target_date=target_date or "",
            )
            summary["milestonesCreated"] += 1
            step(f"Created milestone: {name} ({start_date} → {target_date}).")
        except Exception as err:
            warnings.append(f'Milestone "{name}" failed: {_err_msg(err)}')
            step(f'Milestone "{name}" failed: {_err_msg(err)}', "warn")
            issue_id_matrix.append([])
            continue

        # 4. Create issues for this milestone.
        issue_ids: list = []
        if _cfg(config, "createIssues"):
            for issue in milestone.get("issues") or []:
                title = issue.get("title")
                try:
                    label_ids = await _issue_label_ids(resolve_label, resolve_model_label, task_label_id, issue.get("tshirtSize"))
                    created = await linear.create_issue(
                        api_key,
                        team["id"],
                        project["id"],
                        title,
                        project_milestone_id=created_milestone["id"],
                        description=_with_criteria(issue.get("description"), "Acceptance criteria", issue.get("evaluationCriteria")),
                        priority=issue.get("priority"),
                        label_ids=label_ids if label_ids else None,
                    )
                    issue_ids.append(created["id"])
                    summary["issuesCreated"] += 1
                except Exception as err:
                    issue_ids.append(None)
                    warnings.append(f'Issue "{title}" failed: {_err_msg(err)}')
                    step(f'Issue "{title}" failed: {_err_msg(err)}', "warn")
            ok = len([x for x in issue_ids if x])
            if ok:
                step(f'Created {ok} issue(s) under "{name}".')
        issue_id_matrix.append(issue_ids)

    # 5. Create dependencies (from blocks to). Indices were already validated in
    #    normalize_plan; we re-check the resolved IDs before writing.
    if _cfg(config, "addDependencies") and _cfg(config, "createIssues"):
        for dep in plan.get("dependencies") or []:
            from_id = _safe_at(issue_id_matrix, dep["fromMilestone"], dep["fromIssue"])
            to_id = _safe_at(issue_id_matrix, dep["toMilestone"], dep["toIssue"])
            if not from_id or not to_id or from_id == to_id:
                continue
            try:
                await linear.create_issue_relation(api_key, from_id, to_id, type_="blocks")
                summary["dependenciesCreated"] += 1
            except Exception as err:
                warnings.append(f"Dependency failed: {_err_msg(err)}")
                step(f"Dependency failed: {_err_msg(err)}", "warn")
        if summary["dependenciesCreated"]:
            step(f"Created {summary['dependenciesCreated']} dependency link(s).")

    return summary


async def apply_issues_for_milestones(api_key, *, project, milestones, generated, config, on_step=None):
    """Resume path: create issues for EXISTING milestones that currently have
    none. ``generated`` holds tasks per milestone (same order as ``milestones``);
    matched by index, then by name as a fallback."""
    step = _make_step(on_step)
    warnings: list = []
    summary = {"milestonesCreated": 0, "issuesCreated": 0, "dependenciesCreated": 0, "warnings": warnings, "resumed": True}
    if not _cfg(config, "createIssues"):
        return summary

    team = (await linear.get_project_team(api_key, project["id"]))["team"]
    step(f"Creating tasks for {len(milestones)} milestone(s) (team {team.get('key') or team.get('name')}).")
    task_label_id = await _resolve_task_label_id(api_key, step)
    resolve_label = _make_label_resolver(api_key, step)
    resolve_model_label = _make_model_label_resolver(api_key, step)

    for i in range(len(milestones)):
        milestone = milestones[i]
        m_name = milestone.get("name")
        match = generated[i] if i < len(generated) else None
        if not match:
            wanted = (m_name or "").lower()
            match = next((g for g in generated if (g.get("name") or "").lower() == wanted), None)
        issues = (match.get("issues") if match else None) or []

        # Add the milestone's evaluation criteria to its description (if any).
        if match and match.get("evaluationCriteria"):
            try:
                await linear.update_milestone(
                    api_key,
                    milestone["id"],
                    _with_criteria(milestone.get("description"), "Evaluation criteria", match.get("evaluationCriteria")),
                )
                step(f'Added evaluation criteria to "{m_name}".')
            except Exception as err:
                warnings.append(f"Milestone criteria update failed: {_err_msg(err)}")

        created = 0
        for issue in issues:
            title = issue.get("title")
            try:
                label_ids = await _issue_label_ids(resolve_label, resolve_model_label, task_label_id, issue.get("tshirtSize"))
                await linear.create_issue(
                    api_key,
                    team["id"],
                    project["id"],
                    title,
                    project_milestone_id=milestone["id"],
                    description=_with_criteria(issue.get("description"), "Acceptance criteria", issue.get("evaluationCriteria")),
                    priority=issue.get("priority"),
                    label_ids=label_ids if label_ids else None,
                )
                created += 1
                summary["issuesCreated"] += 1
            except Exception as err:
                warnings.append(f'Issue "{title}" failed: {_err_msg(err)}')
                step(f'Issue "{title}" failed: {_err_msg(err)}', "warn")
        if created:
            step(f'Created {created} task(s) under "{m_name}".')
    return summary


async def apply_aidone(api_key, *, project, on_step=None):
    """Mark a project complete: switch its label to ``aidone`` (replacing others)."""
    step = _make_step(on_step)
    try:
        label = await linear.get_or_create_project_label(api_key, "aidone")
        await linear.set_project_labels(api_key, project["id"], [label["id"]])
        step('Set project label to "aidone".')
    except Exception as err:
        step(f"aidone label failed: {_err_msg(err)}", "warn")


async def apply_aiplanned(api_key, *, project, on_step=None):
    """Mark a project as PLANNED: switch its label to ``aiplanned`` (replacing the
    enrich label). Drops the project out of the planning set and signals the
    coding flow to start working its tasks in dependency order."""
    step = _make_step(on_step)
    try:
        label = await linear.get_or_create_project_label(api_key, "aiplanned")
        await linear.set_project_labels(api_key, project["id"], [label["id"]])
        step('Set project label to "aiplanned".')
    except Exception as err:
        step(f"aiplanned label failed: {_err_msg(err)}", "warn")


async def apply_aifail(api_key, *, project, reason, on_step=None):
    """Handle a project judged NOT viable: append a note to the project
    description and switch its label to ``aifail`` (replacing existing labels so
    it leaves the enrichment set and is not retried)."""
    step = _make_step(on_step)
    note = f"\n\n---\n**AI viability check — not a fit for a software-driven solution.**\nReason: {reason}"
    next_description = f"{project.get('description') or ''}{note}".strip()

    try:
        await linear.update_project_description(api_key, project["id"], next_description)
        step("Added viability note to project description.")
    except Exception as err:
        step(f"Description note failed: {_err_msg(err)}", "warn")

    try:
        label = await linear.get_or_create_project_label(api_key, "aifail")
        await linear.set_project_labels(api_key, project["id"], [label["id"]])
        step('Set project label to "aifail".')
    except Exception as err:
        step(f"Label update failed: {_err_msg(err)}", "warn")

    return {"aifail": True, "reason": reason, "milestonesCreated": 0, "issuesCreated": 0, "dependenciesCreated": 0, "warnings": []}


# --------------------- Coder issue state transitions -------------------- #


async def start_issue(api_key, *, issue_id, on_step=None):
    """Move a coder task to "In Progress" (a ``started`` state) before the agent
    runs. Idempotent: an issue already started/completed is left as-is. Returns
    the resulting state (or None if the team has no started state)."""
    step = _make_step(on_step)
    detail = await linear.get_issue_detail(api_key, issue_id)
    state = detail.get("state")
    type_ = state.get("type") if state else None
    if type_ in ("started", "completed", "canceled"):
        return state or None  # already in/after progress — nothing to do
    target = linear.pick_state_by_type(
        await linear.get_team_states(api_key, detail["team"]["id"]),
        "started",
        "In Progress",
    )
    if not target:
        step('No "In Progress" workflow state on this team; leaving state unchanged.', "warn")
        return state or None
    await linear.update_issue(api_key, issue_id, {"stateId": target["id"]})
    step(f'Moved issue to "{target["name"]}".')
    return target


async def finish_issue(api_key, *, issue_id, outcome, reason=None, on_step=None):
    """Finish a coder task per the agent's verdict: move it to Done (a
    ``completed`` state) and stamp the outcome label on the ISSUE — ``aidone``
    when completed, ``aifail`` when insufficient — creating the label if missing
    and appending it to the issue's existing labels. An insufficient reason is
    posted as a comment."""
    step = _make_step(on_step)
    label_name = "aidone" if outcome == "completed" else "aifail"
    detail = await linear.get_issue_detail(api_key, issue_id)

    # Resolve (create-if-missing) the outcome label, appended to existing labels.
    label_ids = None
    try:
        label = await linear.get_or_create_issue_label(api_key, label_name)
        nodes = (detail.get("labels") or {}).get("nodes") or []
        current = [l["id"] for l in nodes]
        label_ids = list(dict.fromkeys([*current, label["id"]]))
    except Exception as err:
        step(f'Could not resolve "{label_name}" issue label: {_err_msg(err)}', "warn")

    done = linear.pick_state_by_type(
        await linear.get_team_states(api_key, detail["team"]["id"]),
        "completed",
        "Done",
    )
    update_input: dict = {}
    if done:
        update_input["stateId"] = done["id"]
    if label_ids:
        update_input["labelIds"] = label_ids
    if update_input:
        await linear.update_issue(api_key, issue_id, update_input)
    done_label = f'"{done["name"]}"' if done else "(no Done state found)"
    step(f'Marked issue {done_label} + label "{label_name}".')

    # Record the reason a task was judged insufficient (for the human triaging).
    if outcome != "completed" and reason:
        try:
            await linear.create_comment(api_key, issue_id, f"**AI coder — insufficient to complete.**\n\n{reason}")
            step("Posted insufficiency reason as a comment.")
        except Exception as err:
            step(f"Comment failed: {_err_msg(err)}", "warn")
    return {"labelName": label_name, "done": bool(done)}


def _safe_at(matrix, i, j):
    """JS ``matrix[i] && matrix[i][j] ? matrix[i][j] : null`` (indices >= 0)."""
    row = matrix[i] if 0 <= i < len(matrix) else None
    if not row:
        return None
    value = row[j] if 0 <= j < len(row) else None
    return value or None
